File: src/shared/types/resume.ts
/**
 * Resume Parsing Data Contract
 * Output schema produced by Resume Parser module.
 */

import { Skill } from "./skill";

export interface ResumeIdentity {
  full_name: string;
  email: string;
  phone?: string;
  location?: string;
  linkedin_url?: string;
  github_url?: string;
  portfolio_url?: string;
  summary?: string;
}

export interface EducationRecord {
  institution: string;
  degree: string;
  field_of_study: string;
  start_date?: string;
  end_date?: string;
  gpa_or_percentage?: string;
}

export interface ExperienceRecord {
  company: string;
  role: string;
  location?: string;
  start_date?: string;
  end_date?: string;
  is_current?: boolean;
  description?: string;
  key_achievements?: string[];
}

export interface ProjectRecord {
  title: string;
  description: string;
  technologies?: string[];
  github_url?: string;
  live_url?: string;
  highlights?: string[];
}

export interface CertificationRecord {
  name: string;
  issuer: string;
  issue_date?: string;
  expiry_date?: string;
  credential_id?: string;
  credential_url?: string;
}

/**
 * Structured output schema of Resume Parsing
 */
export interface ResumeParseResult {
  id?: string;
  raw_file_name?: string;
  parsed_at: string;
  candidate: ResumeIdentity;
  skills: Skill[];
  education: EducationRecord[];
  experience: ExperienceRecord[];
  projects: ProjectRecord[];
  certifications: CertificationRecord[];
}

File: backend/app/models/resume.py
from typing import List, Optional
from pydantic import BaseModel, Field
from app.models.skill import Skill


class ResumeIdentity(BaseModel):
    full_name: str
    email: str
    phone: Optional[str] = None
    location: Optional[str] = None
    linkedin_url: Optional[str] = None
    github_url: Optional[str] = None
    portfolio_url: Optional[str] = None
    summary: Optional[str] = None


class EducationRecord(BaseModel):
    institution: str
    degree: str
    field_of_study: str
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    gpa_or_percentage: Optional[str] = None


class ExperienceRecord(BaseModel):
    company: str
    role: str
    location: Optional[str] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    is_current: Optional[bool] = None
    description: Optional[str] = None
    key_achievements: Optional[List[str]] = None


class ProjectRecord(BaseModel):
    title: str
    description: str
    technologies: Optional[List[str]] = None
    github_url: Optional[str] = None
    live_url: Optional[str] = None
    highlights: Optional[List[str]] = None


class CertificationRecord(BaseModel):
    name: str
    issuer: str
    issue_date: Optional[str] = None
    expiry_date: Optional[str] = None
    credential_id: Optional[str] = None
    credential_url: Optional[str] = None


class ResumeParseRequest(BaseModel):
    file_name: Optional[str] = None
    raw_text: Optional[str] = None


class ResumeParseResult(BaseModel):
    id: Optional[str] = None
    raw_file_name: Optional[str] = None
    parsed_at: str
    candidate: ResumeIdentity
    skills: List[Skill] = Field(default_factory=list)
    education: List[EducationRecord] = Field(default_factory=list)
    experience: List[ExperienceRecord] = Field(default_factory=list)
    projects: List[ProjectRecord] = Field(default_factory=list)
    certifications: List[CertificationRecord] = Field(default_factory=list)
